### ISSUE TITLE
ModelCheckpoint is not saving top k models

### ISSUE THREAD

**Issue by george-gca** [PyTorchLightning/pytorch-lightning#6117]
## 🐛 Bug

`ModelCheckpoint` is not correctly monitoring metric values.

### To Reproduce

https://colab.research.google.com/drive/1onBmED7dngP_VwFxcFBMsnQi82KbizSk?usp=sharing

### Expected behavior

ModelCheckpoint should save top k models based on `x` metric, but it currently displays `Epoch XXX, step XXX: x was not in top 2` for every epoch.

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.2.0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### Additional context

The documentation doesn't mention how one should set the metric to be used in `ModelCheckpoint`. Tried to use both `x` or `loss` value, but `ModelCheckpoint` shows the same message for both cases. Also, the message should be more clear, saying that `ModelCheckpoint` couldn't find chosen value to monitor instead of saying that it was not in top k, since it displays the same message if I choose to monitor some value that doesn't exist.

**Comment by awaelchli** [comment]
But you never log the metric.

Try this:
```python

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        self.log("x", loss)
```

**Comment by george-gca** [comment]
Ok, maybe the [documentation](https://pytorch-lightning.readthedocs.io/en/stable/extensions/generated/pytorch_lightning.callbacks.ModelCheckpoint.html#pytorch_lightning.callbacks.ModelCheckpoint) should be clear about this.
What about now? I changed the code to call `self.logger.agg_and_log_metrics` and the same issue happens. Also, maybe the message should say it couldn't find the metric.

**Comment by awaelchli** [comment]
Where in the docs did you read that you have to use `self.logger.agg_and_log_metrics`? This bypasses tracking the metric and directly sends it to the logger.

**Comment by george-gca** [comment]
I believe I saw this in an old example in the repository. So, if I want to aggregate my metrics values, I must do it manually, right?

**Comment by awaelchli** [comment]
You reported this issue with 1.2, please see also the corresponding to docs for this version. 
Here is the fixed code:

```python
import os
import pytorch_lightning as pl
import torch
from pytorch_lightning import LightningModule
from torch.utils.data import Dataset, DataLoader


class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.data = torch.randn(num_samples, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


num_samples = 10000

train = RandomDataset(32, num_samples)
train = DataLoader(train, batch_size=32)

val = RandomDataset(32, num_samples)
val = DataLoader(val, batch_size=32)

test = RandomDataset(32, num_samples)
test = DataLoader(test, batch_size=32)


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        for dataset_result in outputs:
            self.logger.agg_and_log_metrics(
                dataset_result, step=self.current_epoch+1)

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        self.log("x", loss)
        return {"x": loss}

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        self.log('fake_test_acc', loss)
        return {"y": loss}

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


if __name__ == "__main__":
    # init model
    model = BoringModel()

    checkpoint_callback = pl.callbacks.ModelCheckpoint(
        dirpath=f'checkpoints',
        mode='min',
        monitor='x',
        verbose=True,
        period=1,
        save_last=True,
        save_top_k=2,
    )
    trainer = pl.Trainer(
        max_epochs=5,
        progress_bar_refresh_rate=20,
        callbacks=[checkpoint_callback]
    )
    trainer.fit(model, train, val)

```

Can you confirm this works?

**Comment by george-gca** [comment]
Changed the code in colab and it is working now. Should I close this bug and open one for the documentation? I checked [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/model_checkpoint.py) and it does not say anything about calling `self.log`. And if I got it right, the docs are automatically generated from the code comments, right?

**Comment by awaelchli** [comment]
Ok good, thanks for the confirmation. Yes we can maybe add a sentence like 

> Every metric logged with `self.log` in LightningModule is a candidate for the monitor key.

No need to close this issue.

> And if I got it right, the docs are automatically generated from the code comments, right?

yes.
Are you interested sending a PR with the doc update?

**Comment by awaelchli** [comment]
You can find further checkpointing tutorials here:

https://pytorch-lightning.readthedocs.io/en/stable/common/weights_loading.html#checkpoint-saving

Maybe it's best if we add a link to that page in ModelCheckoint

**Comment by george-gca** [comment]
Yes, I can send the PR for the doc. What do you think about

> Every metric logged with [self.log](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#log) or [self.log_dict](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#log-dict) in LightningModule is a candidate for the monitor key. For more information, see [checkpoint saving](https://pytorch-lightning.readthedocs.io/en/stable/common/weights_loading.html#checkpoint-saving).

**Comment by george-gca** [comment]
How do I reference the [checkpoint saving](https://pytorch-lightning.readthedocs.io/en/stable/common/weights_loading.html#checkpoint-saving) in the code? I found that to reference [log](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#log) I must do it like

> :meth:`~pytorch_lightning.core.lightning.log`

right?

**Comment by carmocca** [comment]
Yes. Remember you can run `make docs` from the repository root to check that your changes render correctly.

**Comment by carmocca** [comment]
@george-gca still interested in opening the docs PR?

**Comment by george-gca** [comment]
> @george-gca still interested in opening the docs PR?

Yes, but the machine I used to do the coding went down, and with it the Dockerfile I used to create the testing environment. I am waiting for it to come back in the next days. Do you happen to have one Dockerfile that works for this around? I remember I had a few troubles setting mine to work as expected.

**Comment by carmocca** [comment]
A virtual environment with the appropriate dependencies is enough to generate the docs locally with `make docs`. You can also rely on CI to check the docs output
